core/test/solver/gmres.cpp

```diff
@@ -209,6 +209,31 @@ TYPED_TEST(Gmres, CanSetPreconditionerGenerator)
 }
 
 
+TYPED_TEST(Gmres, CanSetCriteriaAgain)
+{
+    using Solver = typename TestFixture::Solver;
+    std::shared_ptr<gko::stop::CriterionFactory> init_crit =
+        gko::stop::Iteration::build().with_max_iters(3u).on(this->exec);
+    auto gmres_factory =
+        Solver::build().with_criteria(init_crit).on(this->exec);
+
+    ASSERT_EQ((gmres_factory->get_parameters().criteria).back(), init_crit);
+
+    auto solver = gmres_factory->generate(this->mtx);
+    std::shared_ptr<gko::stop::CriterionFactory> new_crit =
+        gko::stop::Iteration::build().with_max_iters(5u).on(this->exec);
+
+    solver->set_stop_criterion_factory(new_crit);
```

Comment:
There is one issue here, I think, and I'm not sure what happens then:
All solvers if you look in their constructor, have these lines in their constructor (i.e. factory generate step):
https://github.com/ginkgo-project/ginkgo/blob/587df41c53f38b96b123cc8cbcf8544fbed926b6/include/ginkgo/core/solver/gmres.hpp#L158-L164

What happens in your case then? I'm quite sure that the solver's apply would still be using the wrong criterion, no?
https://github.com/ginkgo-project/ginkgo/blob/500d4a5719ab0b60f76f7f5cbc6f77fd6254f544/core/solver/gmres.cpp#L129-L131

Reply:
It's the same for `krylov_dim` btw.

Reply:
When you use the `set_stop_criterion_factory(fac)`, it should overwrite the value in the current `stop_criterion_factory_` and hence in the `apply_impl` the new generate is called on the stop_crit_fac, right ?

Reply:
My bad, that was a misunderstanding on my part. I now realize you do not change the factory parameters but only the generated solver's. 
I'm not sure I like that inconsistency though, what is the meaning of factories in that case if the solvers deviate from their factories? But indeed you don't have the issue pointed here.

Reply:
I agree that there is a slight inconsistency, but I think it is useful when you dont want to re-generate the whole factory, but just want to reset the parameter when re-running the solver, when the matrix is the same and all other parameters are the same.

Reply:
And by default, you always get whatever you set in the factory build. And these setters can affect only when you call the solver again.